[psram] Allow use of experimental 120MHz octal mode

clydebarrow opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

120MHz PSRAM is available on some ESP32-S3 devices, but is still labelled "experimental" by Espressif for octal PSRAM. This PR enables the use of 120MHz in octal mode, but requires the user to jump through some hoops to make sure they are aware it is not considered stable.

It also adds an `enable_ecc` option for octal mode PSRAM and fixes octal mode PSRAM on Arduino.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4797
## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

clydebarrow committed:
[psram] Allow use of experimental 120MHz octal mode

clydebarrow commented:
To use the changes from this PR as an external component, add the following to your ESPHome configuration YAML file:

```yaml
external_components:
  - source: github://pr#8519
    components: [esp32, psram]
    refresh: 1h
```

---
(Added by the PR bot)

<!-- This comment was generated automatically by a personal bot. -->

clydebarrow commented:
To use the changes from this PR as an external component, add the following to your ESPHome configuration YAML file:

```yaml
external_components:
  - source: github://pr#8519
    components: [esp32, psram]
    refresh: 1h
```

---
(Added by the PR bot)

<!-- This comment was generated automatically by a personal bot. -->

clydebarrow committed:
Clean up logconfig, add ECC option

clydebarrow committed:
Merge branch 'dev' into esp32-experimental

clydebarrow committed:
Fix Arduino debug and octal mode

clydebarrow committed:
syntax

clydebarrow committed:
Braces...

clydebarrow committed:
More brances

clydebarrow committed:
Fix test

Jason2866 commented:
Why enabling ecc? Slows down and does cost PSRAM. In my experience 120Mhz for PSRAM should be used only with IDF >=5.4.1. Earlier minor IDF versions don't have temperature (Re)calibrating.

bdraco commented:
> Why enabling ecc? Slows down and does cost PSRAM. In my experience 120Mhz for PSRAM should be used only with IDF >=5.4.1. Earlier minor IDF versions don't have temperature (Re)calibrating.

Its optional and disabled by default
`cv.Optional(CONF_ENABLE_ECC, default=False): cv.boolean,`

bdraco commented:
Should be good to go once failing test is addressed https://github.com/esphome/esphome/actions/runs/14292270614/job/40055018823?pr=8519

clydebarrow committed:
Remove arduino fix, needs more investigation

clydebarrow commented:
> Should be good to go once failing test is addressed [esphome/esphome/actions/runs/14292270614/job/40055018823?pr=8519](https://github.com/esphome/esphome/actions/runs/14292270614/job/40055018823?pr=8519)

Weirdly that is caused by adding the platformio option needed to make octal PSRAM work on Arduino, but causes build errors in an unrelated component during test (but not when compiling sample code.) That will need more investigation, so for now the Arduino status is unchanged.

If this passes and you're happy, please re-approve, along with the doc PR. Thanks.

Jason2866 commented:
The CI issue is caused by using `platformio/espressif32 @ 5.4.0` which is just wrong
For using IDF it has to be `platformio/espressif32 @ 6.10.0`

bdraco commented:
Totally fine leaving Arduino out of this PR. One small thing—could we update the schema validation to be IDF-only for now?

clydebarrow commented:
> The CI issue is caused by using `platformio/espressif32 @ 5.4.0` which is just wrong

I have no idea what version should be used, and changing that would be out of the scope of this PR, but in any case changing to 6.10.0 does not resolve the CI build problem so it is not the cause of the problem.

clydebarrow committed:
Limit PSRAM to variants that actually support it.
Fix Arduino Octal PSRAM support

clydebarrow commented:
Ok, it seems the test failure was caused by testing on chips that don't support PSRAM. Added validations to restrict that, and removed corresponding tests. Re-instated Arduino octal mode, which now works and passes the tests. So it should be done.

clydebarrow committed:
Merge branch 'dev' into esp32-experimental

clydebarrow committed:
Merge branch 'dev' into esp32-experimental
